### ISSUE TITLE
Extend the follow-up until date calculation for cases and contacts [1]

### ISSUE THREAD

**Issue by Candice-Louw** [hzi-braunschweig/SORMAS-Project#3243]
### Feature Description
As a healthcare worker that has been tasked with following up on cases and contacts (as captured in SORMAS), an amendment to the calculation of the folow-up for cases and contacts is proposed to ensure that the follow-up timeframe is calculated more specifically (accurately).

### Problem Description
Currently, the automatic calculation of follow-up is based on the report date.  This is often inaccurate and values have to be changed manually which becomes quite time consuming.

### Proposed Change
**Cases:** 
A hierarchy is proposed:
- In symptomatic cases, the follow-up should be calculated from the date of onset of symptoms.
- In asymptomatic cases, the calculation should be based on the test acceptance date (**update 06.11.2020: Date sample was collected**).
- when the date of onset of symptoms is entered at a later point of time or newly entered the follow-up time should be changed, but not automatically. There should be a warning "Follow-up should be adjusted due to entry of new data!"
-  When neither the date of onset of symptoms nor the test acceptance date (**update 06.11.2020: Date sample was collected**) is known then the calculation is based on the report date.

**Contacts:**
- The calculation should be based on the date of the last contact.
- If this date is not entered, the report date can be used.
- When the date of last contact is entered at a later point of time or newly entered the follow-up time should be changed, but not automatically. There should be a warning "Follow-up should be adjusted due to entry of new data!"
- 
### Additional Information
Related to: https://github.com/hzi-braunschweig/SORMAS-Project/issues/3100

### Development Specifications
Extend the logic that calculates the follow-up until date of cases and contacts (and validates the follow-up until field) like this:
* If (for a case) a symptom onset date or (for a contact) a last contact date is present, always use it; otherwise:
* If a positive sample has been collected for this case/contact and the sample date time of the EARLIEST sample is BEFORE the report date, use it; otherwise
* Use the report date

**Comment by bernardsilenou** [comment]
@Candice-Louw one significant point you mentioned here is the fact that the users have to manually validate the re-computation of followup when data in the system changes. This was not mentioned in #3100. 

What about followup for event participant? how should it be calculated?
Which variable is sormas that your are referring to as "test acceptance date"?

Please can we have a meeting tomorrow to refine and merge both issues?

**Comment by MateStrysewske** [comment]
@Chinedar Is this fine for Nigeria? I remember that we use the report date on purpose because the onset date is a very unreliable source of information (as in it's based on what the case/contact itself said, so it might be wrong) and, at least for Nigeria, we agreed that it's better to have a slightly longer follow-up than one that is too short.

**Comment by markusmann-vg** [comment]
@Candice-Louw Is this valid for all the SORMAS users or only for x country?
=> users might be very confused about the calculation

**Comment by Chinedar** [comment]
@MateStrysewske This should not apply to Nigeria for now. The calculation of follow-up, determined by case definitions, should rely on the date of report as we currently have it...just like we agreed, it is always better to have a slightly longer follow-up than too short. This ensures that whatever signs and or symptoms that present can be picked up within the defined period of follow-up.

**Comment by MateStrysewske** [comment]
@Chinedar That's what I thought, thanks for your response! We'll definitely not do this change without also adding a configuration for it then.

@markusmann-vg This means that this issue needs to be refined.

**Comment by bernardsilenou** [comment]
@Candice-Louw @kwa20 @GabrielePoggensee  should we need to discuss this further next week?

**Comment by markusmann-vg** [comment]
Test acceptance date is > Case/Contact > New Sample > Date sample was collected

**Comment by bernardsilenou** [comment]
@Candice-Louw @GabrielePoggensee:
This comment is related to this sentence:
_when the date of onset of symptoms is entered at a later point of time or newly entered the follow-up time should be changed, but not automatically. There should be a warning "Follow-up should be adjusted due to entry of new data!"_

- What did the CH users say about a situation that an asymptomatic case is reported, followup and quarantine stated and sample is collected 2 days later? > what should happen to the followup visits before the sample was collected? Should a new quarantine document be issued?

-  Or for the same case, a new sample is collected 10 days after reporting (this often happen in order to confirm that the person is healed)?   should system use only the sample collection date of the oldest sample?
- Or symptom onset date is retrospectively documented to be 7 days before report date, ie after  case has been under followup for 7 days based on report date. Recalculation will imply 14 days of follow up.
- For symptomatic cases, in the time interval between symptom onset date and date of report, no visit could have need done because the visit data is missing by design. Showing this of the follow up chart with a grey box may show as if the contact tracers are not doing their job well.  @GabrielePoggensee  suggested that we should have another kind of icon for this.

**Comment by kwa20** [comment]
@Candice-Louw @bernardsilenou While testing 1.52.0, I noticed that the follow-up for contacts is already calculated by the date of last contact, when entered.

**Comment by kwa20** [comment]
@Candice-Louw @bernardsilenou While testing 1.52.0, I noticed that the follow-up for contacts is already calculated by the date of last contact, when entered.

**Comment by kwa20** [comment]
Users keep requesting this change and I suggest the following to bring the refinement forward:

- Add disease configuration to configure the "most relevant date" by which the follow-up is calculated with (sampleDateTime; onsetDate; reportDate). If configured variable is not entered, use hierarchical dropping to available entries as specified above.
- Add separate configurations for cases and contacts so that a possible necessity to reference different dates is taken into account.

@Candice-Louw Do you think this will be fine for the swiss users?

**Comment by Marko-ilmkreis** [comment]
@kwa20 @Candice-Louw That would be a good feature. 
I would suggest another related change. Currently, when users manually change the follow-up date, an error message appears if the date is below the normal duration of the follow-up. This means that the follow-up can only be extended manually or it has to be cancelled.

It would be useful if the system only checks if the follow-up date is later than the other important dates ( report date, date of last contact, etc.) when a manual entry is made. This also only helps to avoid unintentional incorrect entries (e.g. year 2020 instead of 2021).

**Comment by bernardsilenou** [comment]
@kwa20 @Candice-Louw I have schedule this ticket for further refinement in the meeting on Thursday. It is sensitive and has different view points.

**Comment by Candice-Louw** [comment]
@kwa20 @bernardsilenou - yes for using Symptom onset date, Sample collection date, Report Date as options. 👍

**Comment by kwa20** [comment]
@Marko-ilmkreis I think the problem of being able to reduce the follow-up duration to a timeframe shorter than specified for the disease is that users might not follow-up on people for the appropriate amount of time. The intention for having to cancel it is that users should make a conscious decision and provide a reason for why the follow-up is being canceled. E.g. an early discharge from quarantine might not be reason enough to stop following-up on a person since they might still become symptomatic and eventually transmit the disease. 
@bernardsilenou Do you have something to add to this?

**Comment by bernardsilenou** [comment]
@Marko-ilmkreis  I agree with @kwa20 
- The followup duration is a fixed property of the disease called the incubation period and users are not allowed to change it. 
- If a user decide not to follow up a person for the required incubation period, then they need to document a reason why.

**Comment by Marko-ilmkreis** [comment]
@kwa20 @bernardsilenou I agree that a user should think twice before changing the duration of the follow-up. But perhaps it should be possible to set the follow-up to any date (at least today) if a reason is given for shortening the duration. Maybe with a warning that the normal follow-up period will be shortened by XX days.
In my opinion, this would also have the advantage of showing when the follow-up ended.  So far, you can only see that it has been cancelled, but no when.

In this context, we should also think about deactivating the follow-up automatically for people in contact categories II and III or at least having this configuration option on the server. Alternatively, that these are automatically set to terminated after the end of the follow-up.

According to the German RKI guidelines, follow-up only takes place for category I. For the other categories, only self-monitoring should take place and, in the event of problems, the health authority should be contacted.

**Comment by bernardsilenou** [comment]
* @Marko-ilmkreis One of the goal for adding the contact categories was to do exactly what you said. I though this is what you have been doing.
* @MateStrysewske  Dont we already have a  server configuration to turn on followup for only category I contacts?
* One workaround option to document contact and do not do followup is to set the followup  duration to 0. Good to add such a configuration.
* Another point you mentioned is to be able to know the date the the default followup was interrupted. What about just showing this date when when user click on cancelled follow up or lost to followup, users can also edit the date.

![image](https://user-images.githubusercontent.com/12996724/105412165-a3238180-5c34-11eb-9f1a-e21c0b8db9d5.png)

* It is correct that is should be possible to interrupt followup, eg if person dies, travel abroad, etc In this scenario, the expected end date is not affected and is not defined by the user and it is constant for all entities. 
* One optional point we would like to add a server configuration for the possibility of not calling contact person daily since this may be overwhelming for the health department but rather on a quarantine interval defined the the GE. eh if followup duration is 10, then all automatic task and reminders can be done on days 0, 2,5,8,10  or on 0, 5, 10, day 10 being the most important follow up.

**Comment by MateStrysewske** [comment]
@bernardsilenou No, I don't think we have such a configuration yet

**Comment by kwa20** [comment]
@bernardsilenou changing the follow-up duration configuration to 0 is not an option since it applies to any contacts in the system. Contacts of category 3 are also automatically set to no follow-up. I do however suggest that we could add a configuration to define which categories are taken into consideration for the follow-up calculation, meaning that we either take category 1, 1 and 2, or all 3. This will also allow adjusting to situations where follow-up of all categories is feasible again.
@MateStrysewske I assume this should be handled in another issue?

**Comment by MateStrysewske** [comment]
@kwa20 Exactly!

**Comment by Marko-ilmkreis** [comment]
@kwa20 @bernardsilenou @MateStrysewske I would also suggest that contacts without any category (like travel returns) are also set to no follow up automatically.

**Comment by MateStrysewske** [comment]
Prerequisite: #4241 

Extend the logic that calculates the follow-up until date of cases and contacts (and validates the follow-up until field) like this:
* If (for a case) a symptom onset date or (for a contact) a last contact date is present, always use it; otherwise:
* If a sample has been collected for this case/contact and the sample date time of the EARLIEST sample is BEFORE the report date, use it; otherwise
* Use the report date

**Comment by bernardsilenou** [comment]
I think this issue can solve the need of #4191 but not 100% sure.
When implementing this issue, please have a look at #4191 to validate that the bug is solved.

**Comment by RazvanHodisan** [comment]
Validated on 1.60.0-SNAPSHOT.

**Comment by kwa20** [comment]
@MateStrysewske @bernardsilenou  I noticed that sometimes the follow-up until date contradicts the expected follow-up date which prevents the entity from being saved later on. I suspect that the deposited follow-up duration in the diseaseconfigurations might be the cause. 

tested on: test-de.sormas.netzlink.com

I expect it to be set to 10 for followUpDuration and 14 for caseFollowUpDuration based on the initial calculations. 

<img width="441" alt="followUpError" src="https://user-images.githubusercontent.com/70317594/118687616-845d8200-b805-11eb-8a28-2c4159a4e51e.PNG">
<img width="441" alt="followUpError2" src="https://user-images.githubusercontent.com/70317594/118687625-858eaf00-b805-11eb-8faa-1f6a8f787684.PNG">
<img width="448" alt="followUpError3" src="https://user-images.githubusercontent.com/70317594/118687632-87587280-b805-11eb-9070-f8101fc583b3.PNG">

**Comment by bernardsilenou** [comment]
@Candice-Louw 
I also think it is a bug. Just as you said, the expected followup until logic may be using the hard coded 14 days rather using the value specified on the server

**Comment by MateStrysewske** [comment]
I agree, looks like we don't properly use the values from the disease configurations here.

**Comment by RazvanHodisan** [comment]
Validated on 1.61.0-SNAPSHOT.
